internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,248 @@
+// nextRunIn == 0 the job will be available to run immediately.
+//
+// • jobId must be provided and is the id of the job returned by RegisterJob
+//
+// All options are ignored.
+func (s *Scheduler) UpdateJobNextRun(ctx context.Context, jobId JobId, nextRunIn time.Duration, _ ...Option) error {
+	const op = "scheduler.(Scheduler).UpdateJobNextRun"
+	if jobId == "" {
+		return errors.New(errors.InvalidParameter, op, "missing job id")
+	}
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	_, err = repo.UpdateJobNextRun(ctx, string(jobId), nextRunIn)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+	return nil
+}
+
+// Start begins the scheduling loop that will query the repository for jobs to run and
+// run them in a goroutine.
+func (s *Scheduler) Start() {
+	if s.started.Load() {
+		s.logger.Debug("scheduler already started, skipping")
+		return
+	}
+
+	s.baseContext, s.baseCancel = context.WithCancel(context.Background())
+	go s.start()
+	// TODO (lruch): start go routine to monitor and update progress of running jobs
+	// TODO (lruch): add watchdogs to monitor defunct jobs
+	s.started.Store(true)
+}
+
+func (s *Scheduler) start() {
+	s.logger.Debug("starting scheduling loop")
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-s.baseContext.Done():
+			s.logger.Debug("scheduling loop shutting down")
+			return
+
+		case <-timer.C:
+			s.logger.Debug("waking up to run jobs")
+
+			repo, err := s.jobRepoFn()
+			if err != nil {
+				s.logger.Error("error creating job repo", "error", err)
+				break
+			}
+
+			runs, err := repo.RunJobs(s.baseContext, s.serverId, jobRepo.WithRunJobsLimit(s.runJobsLimit))
+			if err != nil {
+				s.logger.Error("error getting jobs to run from repo", "error", err)
+				break
+			}
+
+			for _, r := range runs {
+				err := s.runJob(r)
+				if err != nil {
+					s.logger.Error("error starting job", "error", err)
+					if _, inner := repo.FailRun(s.baseContext, r.PrivateId); inner != nil {
+						s.logger.Error("error updating failed job run", "error", inner)
+					}
+				}
+			}
+		}
+
+		s.logger.Debug("scheduling loop going back to sleep")
+		timer.Reset(s.runJobsInterval)
+	}
+}
+
+func (s *Scheduler) runJob(r *jobRepo.Run) error {
+	jobId := JobId(r.JobId)
+	j, ok := s.registeredJobs.Load(jobId)
+	job := j.(Job)
+	s.logger.Debug(fmt.Sprintf("starting job run %q for job (%v: %v)", r.PrivateId, jobId, job.Name()))
+	if !ok {
+		return fmt.Errorf("job (%v: %v) not registered on scheduler", jobId, job.Name())
+	}
+
+	s.l.Lock()
+	if _, ok := s.runningJobs[jobId]; ok {
+		s.l.Unlock()
+		return fmt.Errorf("job (%v: %v) is already running", jobId, job.Name())
+	}
+	jobContext, jobCancel := context.WithCancel(context.Background())
+	s.runningJobs[jobId] = runningJob{runId: r.PrivateId, cancelCtx: jobCancel}
+	s.l.Unlock()
+
+	go func() {
+		runErr := job.Run(jobContext)
+		repo, err := s.jobRepoFn()
+		if err != nil {
```

Comment:
Why not make this call before starting the goroutine so the error could be returned and handled by the caller?

Reply:
I wanted to discuss this with you, I was worried about the `repo` sitting around while the job runs as the `job.Run` is a blocking call that blocks for an unknown time